sympy/sets/handlers/functions.py

```diff
@@ -82,7 +89,10 @@ def _set_function(f, x): # noqa:F811
         _end = f(x.end)
 
     if len(sing) == 0:
-        solns = list(solveset(diff(expr, var), var))
```

Comment:
Can this change be reverted now?

Reply:
What test fails without this? All `sets` tests pass for me.

Reply:
This is still necessary, if the differential of `expr` solves to an infinite expression, but doesn't have an infinite amount of singularities. The tests don't actually test this at the moment, but if you apply the patch that you've put, `imageset(x, cos(x), S.Reals)` will hang. I'll add this as another test case now